Update: Add support for parens on left side for-loops (fixes: #8393)

VictorHom opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
Added a fix around supporting parens support for for-in and for-of loops which weren't giving off warnings. I updated no-extra-parens.js with supporting tests.

**Is there anything you'd like reviewers to focus on?**
I wrote a function requiresAfterSpace to determine if a space is required after a node. I didn't know a good way to test this and wasn't sure if what I have is optimal (not sure what I don't know). If you could provide some tips on what a good approach would be to fixing these issues, that would be great. I followed the docs from http://eslint.org/docs/developer-guide/working-with-rules and

not-an-aardvark commented on lib/rules/no-extra-parens.js:
```diff
@@ -256,6 +256,22 @@ module.exports = {
+
+            return rightParenToken && tokenAfterRightParen &&
+                !sourceCode.isSpaceBetweenTokens(rightParenToken, tokenAfterRightParen) &&
+                tokenAfterRightParen.type === "Keyword";
```
Instead of checking whether the rightmost token is a keyword, can you use something like this instead?

```js
const tokenBeforeRightParen = sourceCode.getLastToken(node);
!astUtils.canTokensBeAdjacent(tokenBeforeRightParen, tokenAfterRightParen)
```

The reason a space is needed isn't because the following token is a keyword, it's because the two tokens will combine if the parens are removed. This makes a difference in cases like this:

```js
for ((foo['bar'])of baz);
```

The `]` and `of` would not combine into a single token if the parens were removed, but the autofixer unnecessarily inserts a space at the moment.

VictorHom commented:
@not-an-aardvark thanks

I added the example you provided as a test if that is alright:
`for ((foo['bar'])of baz);`

I updated to check for adjacent nodes as you suggested. Let me know if there are other updates!

kaicataldo commented on tests/lib/rules/no-extra-parens.js:
```diff
@@ -1005,3 +1025,5 @@ ruleTester.run("no-extra-parens", rule, {
     ]
 });
+
+
```
Nit: Extra newline here

not-an-aardvark replied:
Maybe we should enforce this with `no-multiple-empty-lines` on the codebase.

kaicataldo replied:
Yes, agreed!

kaicataldo commented on lib/rules/no-extra-parens.js:
```diff
@@ -256,6 +256,23 @@ module.exports = {
+         * @returns {boolean} `true` if a space should be inserted after the node
+         * @private
+         */
+        function requiresAfterSpace(node) {
```
Nitpick: `requiresTrailingSpace` might be a little clearer for this function name

VictorHom commented:
@kaicataldo @not-an-aardvark addressed the comments.
